ibm/mas_devops/roles/cp4d_install_services/tasks/cpd40/cpd-install-services.yml

```diff
@@ -24,6 +24,12 @@
     msg: "cpd_block_storage_class property has not been set but it is required to install IBM Watson Discovery"
   with_items: "{{ cpd_services }}"
 
+# Note: this role is always called together with cpd_install and it is because of this that
```

Comment:
That's not true, at least to best of my understanding.  

The reason they are seperate is so that a user can independently call `cpd_install_services` and `cpd_install` roles ... e.g. today you could `cpd_install` and `cpd_install_services` (db2) .. then tomorrow you could `cpd_install_services` (watson studio) ... we can't rely on variables that are only set in another role being inherited. 

However, lookign at the code I think the interpretation of how it works is wrong in general.  The reason it works is that all of our playbooks that call this role explicitly define the `cpd_meta_namespace` variable ... it's not due to any link to the cpd_install role.  

tldr; I think this comment block is wrong, but I could be mistaken :)

Reply:
> The reason it works is that all of our playbooks that call this role explicitly define the cpd_meta_namespace variable .

It's not I see today... here lite manage for example: https://github.com/ibm-mas/ansible-devops/blob/master/ibm/mas_devops/playbooks/lite-manage-roks.yml#L46-L49 - and even if we did, for cpd 4.0 cpd_meta_namespace would not be enough, as we have two namespaces

> we can't rely on variables that are only set in another role being inherited; if that is the case it's a major bug.

The cpd 4.0 vars were only loaded here: https://github.com/ibm-mas/ansible-devops/blob/master/ibm/mas_devops/roles/cp4d_install/tasks/cpd40/cpd-install.yml#L5 and there is nothing else that pass such values to the other role (called, usually, right after this one by our cpd playbooks)

So the comment is more to raise such TODO to fix, as I took some time to figure it out, instead of fixing that now and then need to retest everything again - we can do that in a subsequent patch (and then more tests), so we do not hold this stuff for so long. Makes sense?

Reply:
> It's not I see today... here lite manage for example: https://github.com/ibm-mas/ansible-devops/blob/master/ibm/mas_devops/playbooks/lite-manage-roks.yml#L46-L49

The way that works is through variable priority:

1. lite-manage-roks imports the install-services-db2 playbook and doesn't set any value for `cpd_meta_namespace` in it's vars, if it did this would override anything set in the playbook we import, or in the defaults built into the role
2. the imported install-services-db2 playbook sets a value for `cpd_meta_namespace` (based on `CPD_NAMESPACE` environment variable, defaulting to `cpd-meta-ops` if that env var is not set), so that's what we use, which would override any defaults built into the role (if we had one), but would itself be overridden if we were to set a value in the lite-manage-roks playbook import.

Because there is no setting in defaults/main.yml we are effectively declaring `cpd_meta_namespace` as a required variable that must be passed in whenever running this role.  There's no connection to the cpd-install role other than that they both happen to be called in the same playbook in our sample playbooks.

Reply:
1 and 2 above are correct, yes, but let's continue the flow:

1. lite-manage-roks imports install-services-db2 and do not set the cpd_meta_namespace var
2. install-services-db2 defaults cpd_meta_namespace to either the environment var or to cpd-meta-ops if not set, ok.
3. cpd-install role starts. If cpd version is 35, we use the cpd_meta_namespace defaulted in the playbook, no problem
3.1 if cpd version is 4, we load vars [here](https://github.com/ibm-mas/ansible-devops/blob/master/ibm/mas_devops/roles/cp4d_install/tasks/cpd40/cpd-install.yml#L5), and now cpd_meta_namespace is overwritten AND a new var appears: `cpd_services_namespace` - in other words we are ignoring what have been passed, but let's move forward
4. cpd-isntall-services role starts... if cpd 35, we will continue using the var defaulted in playbook. But, if it is 4.0, we are using the vars from 3.1... otherwise, `cpd_services_namespace` would not even exist as nobody set that besides the cpd-install